wicker/core/datasets.py

```diff
@@ -303,3 +313,248 @@ def dataset_size(self) -> int:
+            dataset_partition_name (str): Partition name.
+            table_name (str): Name of the table in database to load.
+            connection_parameters (Optional[Dict[str, str]], optional):
+                Parameters to connect to the database. Defaults to None.
+            connection (Any, optional): Established connection to the database. Defaults to None.
+            schema (Optional[DatasetSchema], optional): Expected DatasetSchema. Defaults to None.
+            columns_to_load (Optional[List[str]], optional): Name of columns to load. Defaults to None.
+            optional_condition (str, optional): Query condition to specify the dataset. Defaults to "".
+            primary_keys (List[str], optional): Primary key of the dataset.
+                Defaults to ["scene_id", "frame_idx"].
+        """
+        self._arrow_table: Optional[pyarrow.Table] = None
+
+        self._dataset_id = DatasetID(name=dataset_name, version=dataset_version)
+        self._partition = DatasetPartition(dataset_id=self._dataset_id, partition=dataset_partition_name)
+        self._schema = schema
+        self._table_name = table_name
+        self._columns_to_load = columns_to_load
+        self._primary_keys = primary_keys
+        self._connection_parameters = connection_parameters
+        self._optional_condition = optional_condition
+        self._connection = connection
+        self._dataset_definition = DatasetDefinition(
+            self._dataset_id,
+            schema=self.schema(),
+        )
+
+    @property
+    def connection(self):
+        """Returns a connection with the database.
+            It tries to connect if no connection exists.
+
+        Raises:
+            ValueError: Error if both connection and connection_parameters are None.
+
+        Returns:
+            _type_: Connection to the database.
+        """
+        if self._connection is None:
+            if self._connection_parameters is not None:
+                self._connection = snowflake.connector.connect(**self._connection_parameters)
+            else:
+                raise ValueError(
+                    "Invalid input: Both 'connection' and 'connection_parameters' cannot be None.",
+                    "At least one of them must be provided.",
+                )
+        return self._connection
+
+    def schema(self) -> DatasetSchema:
+        """Returns schema of the dataset. It can be inputted by end-user to ensure the expected schema.
+
+        Returns:
+            DatasetSchema: Schema of the dataset.
+        """
+        if self._schema is None:
+            schema_table = self._get_schema_from_database()
+            schema_fields = self._get_schema_fields(schema_table)
+            self._schema = DatasetSchema(schema_fields, self._primary_keys)
+        return self._schema
+
+    def arrow_table(self) -> pyarrow.Table:
+        """Returns a table of the dataset as pyarrow table.
```

Comment:
General speaking, I am not sure whether we should wrap the sf database table as a pyarrow table.

1.  loading performance, every data loading process need a copy of this dataset , aka a connection to the snowflake and a remote arrow table loading. Even Snowflake is designed as a OLAP instead of OLTP, but theoretically, the DB service overhead should be much slower than a direct S3 connection.
2. Data volume, pyarrow loading from local file(downloaded from s3 although) utilize the mmap, theoretically support larger data volume than the memory, but your approach needs loading into memory , no local file based cache.
The design principle of wicker is utilize a ETL stage(data dumping piple) to transform the format ,here parquet+column files(aka, heavy duty field, should only for the image like large size field) for more IO friendly for the gpu instances to shorten the loading time to save money.
Based on the above, I would request for changes for your PR and let us have f2f discuss.